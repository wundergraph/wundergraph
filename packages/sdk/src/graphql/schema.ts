import {
	GraphQLSchema,
	Kind,
	OperationTypeDefinitionNode,
	OperationTypeNode,
	parse,
	print,
	printSchema,
	SchemaDefinitionNode,
	StringValueNode,
	visit,
} from 'graphql';
import { GraphQLIntrospection } from '../definition';

export const cleanupSchema = (schema: GraphQLSchema, introspection: GraphQLIntrospection): string => {
	const customFloatScalars = introspection.customFloatScalars || [];
	const customIntScalars = introspection.customIntScalars || [];

	const printed = printSchema(schema);
	const ast = parse(printed);
	const queryTypeName = schema.getQueryType()?.name;
	const mutationTypeName = schema.getMutationType()?.name;
	const subscriptionTypeName = schema.getSubscriptionType()?.name;
	const cleanAst = visit(ast, {
		SchemaDefinition: (node) => {
			const operationTypes: OperationTypeDefinitionNode[] = [];
			if (queryTypeName) {
				operationTypes.push({
					kind: Kind.OPERATION_TYPE_DEFINITION,
					operation: OperationTypeNode.QUERY,
					type: {
						kind: Kind.NAMED_TYPE,
						name: {
							kind: Kind.NAME,
							value: 'Query',
						},
					},
				});
			}
			if (mutationTypeName) {
				operationTypes.push({
					kind: Kind.OPERATION_TYPE_DEFINITION,
					operation: OperationTypeNode.MUTATION,
					type: {
						kind: Kind.NAMED_TYPE,
						name: {
							kind: Kind.NAME,
							value: 'Mutation',
						},
					},
				});
			}
			if (subscriptionTypeName) {
				operationTypes.push({
					kind: Kind.OPERATION_TYPE_DEFINITION,
					operation: OperationTypeNode.SUBSCRIPTION,
					type: {
						kind: Kind.NAMED_TYPE,
						name: {
							kind: Kind.NAME,
							value: 'Subscription',
						},
					},
				});
			}
			const schemaNode: SchemaDefinitionNode = {
				...node,
				operationTypes,
			};
			return schemaNode;
		},
		ObjectTypeDefinition: (node) => {
			if (node.name.value === queryTypeName) {
				return {
					...node,
					name: {
						...node.name,
						value: 'Query',
					},
				};
			}

			if (node.name.value === mutationTypeName) {
				return {
					...node,
					name: {
						...node.name,
						value: 'Mutation',
					},
				};
			}

			if (node.name.value === subscriptionTypeName) {
				return {
					...node,
					name: {
						...node.name,
						value: 'Subscription',
					},
				};
			}

			switch (node.name.value) {
				case '_Service':
				case 'Entity':
					return null;
			}
			if (node.name.value.startsWith('__')) {
				return null;
			}
		},
		UnionTypeDefinition: (node) => {
			switch (node.name.value) {
				case '_Entity':
					return null;
			}
			if (node.name.value.startsWith('__')) {
				return null;
			}
		},
		ScalarTypeDefinition: (node) => {
			switch (node.name.value) {
				case '_Any':
				case '_FieldSet':
					return null;
			}
			if (node.name.value.startsWith('__')) {
				return null;
			}
			if (customFloatScalars.includes(node.name.value)) {
				return null;
			}
			if (customIntScalars.includes(node.name.value)) {
				return null;
			}
		},
		NamedType: (node) => {
			if (customFloatScalars.includes(node.name.value)) {
				return {
					...node,
					name: {
						...node.name,
						value: 'Float',
					},
				};
			}
			if (customIntScalars.includes(node.name.value)) {
				return {
					...node,
					name: {
						...node.name,
						value: 'Int',
					},
				};
			}
		},
		DirectiveDefinition: (node) => {
			switch (node.name.value) {
				case 'key':
				case 'extends':
				case 'external':
				case 'requires':
				case 'provides':
					return null;
			}
		},
		FieldDefinition: {
			enter: (node) => {
				switch (node.name.value) {
					case '_entities':
					case '_service':
						return null;
				}
				if (node.name.value.startsWith('__')) {
					return null;
				}
				if (node.description) {
					if (node.description.value === '') {
						return {
							...node,
							description: undefined,
						};
					}
					if (node.description.block === true) {
						const description: StringValueNode = {
							...node.description,
							block: false,
						};
						return {
							...node,
							description,
						};
					}
				}
			},
		},
		InputValueDefinition: {
			enter: (node) => {
				if (node.description) {
					if (node.description.value === '') {
						return {
							...node,
							description: undefined,
						};
					}
					if (node.description.block === true) {
						const description: StringValueNode = {
							...node.description,
							block: false,
						};
						return {
							...node,
							description,
						};
					}
				}
			},
		},
	});

	return print(cleanAst);
};
