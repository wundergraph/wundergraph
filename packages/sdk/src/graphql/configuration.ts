import {
	buildASTSchema,
	DocumentNode,
	GraphQLSchema,
	OperationDefinitionNode,
	parse,
	SelectionSetNode,
	visit,
} from 'graphql';
import {
	ArgumentConfiguration,
	ArgumentRenderConfiguration,
	ArgumentSource,
	FieldConfiguration,
	TypeConfiguration,
	TypeField,
} from '@wundergraph/protobuf';
import { ArgumentReplacement } from '../transformations/transformSchema';
import { TypeNode } from 'graphql/language/ast';
import { Kind } from 'graphql/language/kinds';

const DefaultJsonType = 'JSON';

export interface GraphQLConfiguration {
	RootNodes: TypeField[];
	ChildNodes: TypeField[];
	Fields: FieldConfiguration[];
	Types: TypeConfiguration[];
}

export const configuration = (
	schema: DocumentNode,
	customJsonScalars: string[] = [],
	serviceSDL?: DocumentNode,
	argumentReplacements?: ArgumentReplacement[]
): GraphQLConfiguration => {
	const config: GraphQLConfiguration = {
		RootNodes: [],
		ChildNodes: [],
		Fields: [],
		Types: [],
	};
	const replacements = argumentReplacements || [];
	if (serviceSDL !== undefined) {
		visitSchema(serviceSDL, config, customJsonScalars, true, replacements);
	} else {
		visitSchema(schema, config, customJsonScalars, false, replacements);
	}
	return config;
};

interface JsonTypeField {
	typeName: string;
	fieldName: string;
}

const visitSchema = (
	schema: DocumentNode,
	config: GraphQLConfiguration,
	customJsonScalars: string[],
	isFederation: boolean,
	argumentReplacements: ArgumentReplacement[]
) => {
	let typeName: undefined | string;
	let fieldName: undefined | string;
	let isExtensionType = false;
	let hasExtensionDirective = false;
	let isEntity = false;
	let isExternalField = false;
	let entityFields: string[] = [];
	let jsonFields: JsonTypeField[] = [];

	const jsonScalars = [DefaultJsonType];
	jsonScalars.push(...customJsonScalars);

	const RootNodeNames = rootNodeNames(schema, isFederation);
	const isNodeRoot = (typeName: string) => {
		return RootNodeNames.includes(typeName);
	};

	visit(schema, {
		ObjectTypeDefinition: {
			enter: (node) => {
				typeName = node.name.value;
				isExtensionType = false;
				isEntity = false;
			},
			leave: () => {
				typeName = undefined;
				isExtensionType = false;
				hasExtensionDirective = false;
				entityFields = [];
				isEntity = false;
			},
		},
		InterfaceTypeDefinition: {
			enter: (node) => {
				typeName = node.name.value;
				isExtensionType = false;
				isEntity = false;
			},
			leave: () => {
				typeName = undefined;
				isExtensionType = false;
				hasExtensionDirective = false;
				entityFields = [];
				isEntity = false;
			},
		},
		ObjectTypeExtension: {
			enter: (node) => {
				typeName = node.name.value;
				isExtensionType = true;
				isEntity = false;
			},
			leave: () => {
				typeName = undefined;
				isExtensionType = false;
				hasExtensionDirective = false;
				entityFields = [];
			},
		},
		InterfaceTypeExtension: {
			enter: (node) => {
				typeName = node.name.value;
				isExtensionType = true;
				isEntity = false;
			},
			leave: () => {
				typeName = undefined;
				isExtensionType = false;
				hasExtensionDirective = false;
				entityFields = [];
			},
		},
		Directive: {
			enter: (node) => {
				switch (node.name.value) {
					case 'extends':
						hasExtensionDirective = true;
						return;
					case 'key':
						isEntity = true;
						if (!node.arguments) {
							return;
						}
						const fields = node.arguments.find((arg) => arg.name.value === 'fields');
						if (!fields) {
							return;
						}
						if (fields.value.kind !== 'StringValue') {
							return;
						}
						const fieldsValue = fields.value.value;
						const fieldsSelection = parseSelectionSet('{ ' + fieldsValue + ' }');
						fieldsSelection.selections.forEach((s) => {
							if (s.kind !== 'Field') {
								return;
							}
							entityFields.push(s.name.value);
						});
						return;
					case 'external':
						isExternalField = true;
				}
			},
		},
		FieldDefinition: {
			enter: (node) => {
				fieldName = node.name.value;

				if (isJsonField(node.type, jsonScalars)) {
					jsonFields.push({ typeName: typeName!, fieldName: fieldName! });
				}
			},
			leave: () => {
				if (typeName === undefined || fieldName === undefined) {
					return;
				}
				const isRoot = isNodeRoot(typeName);
				if (isRoot) {
					addTypeField(config.RootNodes, typeName, fieldName);
				}

				const isExtension = isExtensionType || hasExtensionDirective;
				const isFederationRootNode = isExtension && isEntity && !isExternalField;
				const isEntityField = entityFields.find((f) => f === fieldName) !== undefined;

				if (isEntity && !isExternalField) {
					addTypeField(config.RootNodes, typeName, fieldName);
				}

				if (isFederationRootNode) {
					addTypeField(config.RootNodes, typeName, fieldName);
					addRequiredFields(typeName, fieldName, config, entityFields);
				}

				if (!isRoot && !isFederationRootNode && !isExternalField) {
					addTypeField(config.ChildNodes, typeName, fieldName);
				}

				if (isExternalField && isEntityField) {
					addTypeField(config.ChildNodes, typeName, fieldName);
				}

				if (isEntity && !isEntityField && !isExternalField && !isFederationRootNode) {
					addRequiredFields(typeName, fieldName, config, entityFields);
				}

				fieldName = undefined;
				isExternalField = false;
			},
		},
		InputValueDefinition: {
			enter: (node) => {
				if (!fieldName || !typeName) {
					return;
				}
				addFieldArgument(typeName, fieldName, node.name.value, config, argumentReplacements);
			},
		},
	});

	addJsonFieldConfigurations(config, jsonFields);
};

const parseSelectionSet = (selectionSet: string): SelectionSetNode => {
	const query = parse(selectionSet).definitions[0] as OperationDefinitionNode;
	return query.selectionSet;
};

const rootNodeNames = (schema: DocumentNode, isFederation: boolean): string[] => {
	const rootTypes = new Set<string>();
	visit(schema, {
		SchemaDefinition: {
			enter: (node) => {
				node.operationTypes.forEach((operationType) => {
					rootTypes.add(operationType.type.name.value);
				});
			},
		},
		ObjectTypeDefinition: {
			enter: (node) => {
				switch (node.name.value) {
					case 'Query':
					case 'Mutation':
					case 'Subscription':
						rootTypes.add(node.name.value);
				}
			},
		},
		ObjectTypeExtension: {
			enter: (node) => {
				if (!isFederation) {
					return;
				}
				switch (node.name.value) {
					case 'Query':
					case 'Mutation':
					case 'Subscription':
						rootTypes.add(node.name.value);
				}
			},
		},
	});

	return Array.from(rootTypes.values());
};

export const isRootType = (typeName: string, schema: GraphQLSchema): boolean => {
	const queryType = schema.getQueryType();
	if (queryType && queryType.astNode && queryType.astNode.name.value === typeName) {
		return true;
	}
	const mutationType = schema.getMutationType();
	if (mutationType && mutationType.astNode && mutationType.astNode.name.value === typeName) {
		return true;
	}
	const subscriptionType = schema.getSubscriptionType();
	if (subscriptionType && subscriptionType.astNode && subscriptionType.astNode.name.value === typeName) {
		return true;
	}
	const typeDefinition = schema.getType(typeName);
	if (
		typeDefinition === undefined ||
		typeDefinition === null ||
		typeDefinition.astNode === undefined ||
		typeDefinition.astNode === null
	) {
		return false;
	}
	return false;
};

const addTypeField = (typeFields: TypeField[], typeName: string, fieldName: string) => {
	const i = typeFields.findIndex((t) => t.typeName === typeName);
	if (i !== -1) {
		addField(typeFields[i], fieldName);
		return;
	}
	const typeField: TypeField = {
		typeName: typeName,
		fieldNames: [],
	};
	addField(typeField, fieldName);
	typeFields.push(typeField);
};

const addField = (typeField: TypeField, field: string) => {
	const i = typeField.fieldNames.findIndex((f) => f === field);
	if (i !== -1) {
		return;
	}
	typeField.fieldNames.push(field);
};

const addFieldArgument = (
	typeName: string,
	fieldName: string,
	argName: string,
	config: GraphQLConfiguration,
	argumentReplacements: ArgumentReplacement[]
) => {
	const replacement = argumentReplacements.find((argument) => {
		return argument.argName == argName && argument.fieldName == fieldName && argument.typeName == typeName;
	});
	const arg: ArgumentConfiguration = {
		name: argName,
		sourceType: ArgumentSource.FIELD_ARGUMENT,
		sourcePath: [],
		renderConfiguration: ArgumentRenderConfiguration.RENDER_ARGUMENT_DEFAULT,
		renameTypeTo: replacement?.renameTypeTo || '',
	};
	let field: FieldConfiguration | undefined = findField(config.Fields, typeName, fieldName);
	if (!field) {
		config.Fields.push({
			typeName: typeName,
			fieldName: fieldName,
			argumentsConfiguration: [arg],
			disableDefaultFieldMapping: false,
			path: [],
			requiresFields: [],
			unescapeResponseJson: false,
		});
		return;
	}
	if (!field.argumentsConfiguration) {
		field.argumentsConfiguration = [arg];
		return;
	}
	const i = field.argumentsConfiguration.findIndex((a) => a.name === argName);
	if (i !== -1) {
		field.argumentsConfiguration[i] = arg;
		return;
	}
	field.argumentsConfiguration.push(arg);
};

const addRequiredFields = (
	typeName: string,
	fieldName: string,
	config: GraphQLConfiguration,
	requiredFieldNames: string[]
) => {
	requiredFieldNames.forEach((f) => {
		addRequiredField(typeName, fieldName, config, f);
	});
};

const addRequiredField = (
	typeName: string,
	fieldName: string,
	config: GraphQLConfiguration,
	requiredFieldName: string
) => {
	let field: FieldConfiguration | undefined = findField(config.Fields, typeName, fieldName);
	if (!field) {
		config.Fields.push({
			typeName: typeName,
			fieldName: fieldName,
			requiresFields: [requiredFieldName],
			argumentsConfiguration: [],
			path: [],
			disableDefaultFieldMapping: false,
			unescapeResponseJson: false,
		});
		return;
	}
	if (!field.requiresFields) {
		field.requiresFields = [requiredFieldName];
		return;
	}
	const exists = field.requiresFields.find((f) => f === requiredFieldName) !== undefined;
	if (exists) {
		return;
	}
	field.requiresFields.push(requiredFieldName);
};

const addJsonFieldConfigurations = (config: GraphQLConfiguration, jsonFields: JsonTypeField[]) => {
	for (const jsonField of jsonFields) {
		let field: FieldConfiguration | undefined = findField(config.Fields, jsonField.typeName, jsonField.fieldName);

		if (field) {
			field.unescapeResponseJson = true;
		} else {
			config.Fields.push({
				typeName: jsonField.typeName,
				fieldName: jsonField.fieldName,
				argumentsConfiguration: [],
				disableDefaultFieldMapping: false,
				path: [],
				requiresFields: [],
				unescapeResponseJson: true,
			});
		}
	}
};

const findField = (fields: FieldConfiguration[], typeName: string, fieldName: string) => {
	return fields.find((f) => f.typeName === typeName && f.fieldName === fieldName);
};

const isJsonField = (type: TypeNode, jsonScalars: string[]) => {
	const namedTypeName = resolveNamedTypeName(type);

	for (const jsonType of jsonScalars) {
		if (namedTypeName === jsonType) {
			return true;
		}
	}

	return false;
};

const resolveNamedTypeName = (type: TypeNode): string => {
	switch (type.kind) {
		case Kind.NON_NULL_TYPE:
			return resolveNamedTypeName(type.type);
		case Kind.LIST_TYPE:
			return resolveNamedTypeName(type.type);
		default:
			return type.name.value;
	}
};
