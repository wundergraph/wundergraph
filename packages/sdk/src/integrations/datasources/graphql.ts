import type { GraphQLIntrospection } from '../../definition';
import { defineDatasource } from './define-datasource';

export interface GraphQLDatasourceOptions extends Omit<GraphQLIntrospection, 'apiNamespace'> {
	namespace?: string;
}

export const graphql = defineDatasource<GraphQLDatasourceOptions>((config) => {
	const { namespace, ...introspectionConfig } = config;
	return {
		name: 'graphql',
		hooks: {
			'config:setup': async (options) => {
				const { introspect } = await import('../../definition');
				options.addApi(
					introspect.graphql({
						apiNamespace: namespace,
						...introspectionConfig,
					})
				);
			},
		},
	};
});
