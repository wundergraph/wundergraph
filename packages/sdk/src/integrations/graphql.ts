import type { GraphQLIntrospection } from '../definition';
import { defineDatasource } from './define-datasource';

export interface GraphQLDatasourceOptions extends Omit<GraphQLIntrospection, 'apiNamespace'> {
	namespace: string;
}

export const graphql = defineDatasource<GraphQLDatasourceOptions>((config) => {
	return {
		name: 'graphql',
		hooks: {
			'config:setup': async (options) => {
				const { introspect } = await import('../definition');
				options.addApi(
					introspect.graphql({
						apiNamespace: config.namespace,
						...config,
					})
				);
			},
		},
	};
});
