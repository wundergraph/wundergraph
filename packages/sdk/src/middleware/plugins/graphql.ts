import { FastifyPluginAsync } from 'fastify';
import fp from 'fastify-plugin';
import { GraphQLSchema } from 'graphql';
import {
	getGraphQLParameters,
	processRequest,
	renderGraphiQL,
	sendResult,
	shouldRenderGraphiQL,
	ExecutionContext as HelixExecutionContext,
} from 'graphql-helix';
import { BaseContext } from '../server';

export interface GraphQLServerConfig {
	serverName: string;
	schema: GraphQLSchema | Promise<GraphQLSchema>;
	apiNamespace?: string;
	enableGraphQLEndpoint?: boolean;
	routeUrl: string;
	// use baseContext to pass global data to the graphql context
	baseContext?: Object | Promise<Object>;
	// use the contextFactory to create a context for each request
	// the callback needs to be called with your object
	contextFactory?: (callback: (ctx: Object) => void) => Promise<void>;
	// by default, root fields will be prefixed with the api namespace
	// use skipRenameRootFields to skip this behaviour, e.g. if a field is already prefixed
	skipRenameRootFields?: string[];
	// implement resolverFactory to create a custom resolver
	customResolverFactory?: (executionContext: HelixExecutionContext & ExecutionContext) => Promise<{}>;
}

interface ExecutionContext {
	wundergraph: BaseContext;
}

const FastifyGraphQLPlugin: FastifyPluginAsync<GraphQLServerConfig> = async (fastify, config) => {
	const schema = await config.schema;
	const baseContext = await config.baseContext;

	fastify.route({
		method: ['GET', 'POST'],
		url: config.routeUrl,
		async handler(req, res) {
			const request = {
				body: req.body,
				headers: req.headers,
				method: req.method,
				query: req.query,
			};

			const pluginLogger = req.ctx.log.child({ server: config.serverName, plugin: 'graphql' });

			if (config.enableGraphQLEndpoint && shouldRenderGraphiQL(request)) {
				res.type('text/html');
				res.send(
					renderGraphiQL({
						endpoint: config.routeUrl,
					})
				);
			} else {
				const request = {
					body: req.body,
					headers: req.headers,
					method: req.method,
					query: req.query,
				};
				const { operationName, query, variables } = getGraphQLParameters(request);

				if (config.contextFactory) {
					await config.contextFactory(async (ctx) => {
						const result = await processRequest<ExecutionContext>({
							operationName,
							query,
							variables,
							request,
							schema,
							// @ts-ignore
							rootValueFactory: config.customResolverFactory,
							contextFactory: (): ExecutionContext => ({
								...baseContext,
								...ctx,
								wundergraph: {
									...req.ctx,
									log: pluginLogger,
								},
							}),
						});

						await sendResult(result, res.raw);
						res.sent = true;
					});
					return;
				}

				const result = await processRequest<ExecutionContext>({
					operationName,
					query,
					variables,
					request,
					schema,
					// @ts-ignore
					rootValueFactory: config.customResolverFactory,
					contextFactory: (): ExecutionContext => ({
						...baseContext,
						wundergraph: {
							...req.ctx,
							log: pluginLogger,
						},
					}),
				});

				await sendResult(result, res.raw);

				// skip async/await handling because we already sent the result with the raw http socket
				res.sent = true;
			}
		},
	});
};

export default fp(FastifyGraphQLPlugin, '3.x');
