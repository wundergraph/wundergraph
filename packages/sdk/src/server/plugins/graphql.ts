import { FastifyPluginAsync } from 'fastify';
import { GraphQLSchema } from 'graphql';
import {
	getGraphQLParameters,
	processRequest,
	renderGraphiQL,
	sendResult,
	shouldRenderGraphiQL,
	ExecutionContext as HelixExecutionContext,
} from 'graphql-helix';
import type { BaseRequestContext } from '../types';

export interface GraphQLServerConfig {
	serverName: string;
	schema: GraphQLSchema | Promise<GraphQLSchema>;
	apiNamespace?: string;
	enableGraphQLEndpoint?: boolean;
	routeUrl: string;
	// use baseContext to pass global data to the graphql context
	baseContext?: Object | Promise<Object>;
	// use the contextFactory to create a context for each request
	// the callback needs to be called with your object
	contextFactory?: (callback: (ctx: Object) => void) => Promise<void>;
	// by default, root fields will be prefixed with the api namespace
	// use skipRenameRootFields to skip this behaviour, e.g. if a field is already prefixed
	skipRenameRootFields?: string[];
	// implement resolverFactory to create a custom resolver
	customResolverFactory?: (executionContext: HelixExecutionContext & ExecutionContext) => Promise<{}>;
}

interface ExecutionContext {
	wundergraph: BaseRequestContext;
}

const FastifyGraphQLPlugin: FastifyPluginAsync<GraphQLServerConfig> = async (fastify, config) => {
	const schema = await config.schema;
	const baseContext = await config.baseContext;

	fastify.route({
		method: ['GET', 'POST'],
		url: config.routeUrl,
		async handler(req, reply) {
			const request = {
				body: req.body,
				headers: req.headers,
				method: req.method,
				query: req.query,
			};

			const pluginLogger = req.ctx.log.withFields({ server: config.serverName, plugin: 'graphql' });

			if (config.enableGraphQLEndpoint && shouldRenderGraphiQL(request)) {
				reply.type('text/html');
				reply.send(
					renderGraphiQL({
						endpoint: config.routeUrl,
					})
				);
			} else {
				// https://www.fastify.io/docs/latest/Reference/Reply/#hijack
				// We hand over the response handling to "graphql-helix"
				// No fastify hooks are called for the response.
				reply.hijack();

				const request = {
					body: req.body,
					headers: req.headers,
					method: req.method,
					query: req.query,
				};
				const { operationName, query, variables } = getGraphQLParameters(request);

				if (config.contextFactory) {
					await config.contextFactory(async (ctx) => {
						const result = await processRequest<ExecutionContext>({
							operationName,
							query,
							variables,
							request,
							schema,
							// @ts-ignore
							rootValueFactory: config.customResolverFactory,
							contextFactory: (): ExecutionContext => ({
								...baseContext,
								...ctx,
								wundergraph: {
									...req.ctx,
									log: pluginLogger,
								},
							}),
						});

						await sendResult(result, reply.raw);
					});
					return;
				}

				const result = await processRequest<ExecutionContext>({
					operationName,
					query,
					variables,
					request,
					schema,
					// @ts-ignore
					rootValueFactory: config.customResolverFactory,
					contextFactory: (): ExecutionContext => ({
						...baseContext,
						wundergraph: {
							...req.ctx,
							log: pluginLogger,
						},
					}),
				});

				await sendResult(result, reply.raw);
			}
		},
	});
};

export default FastifyGraphQLPlugin;
