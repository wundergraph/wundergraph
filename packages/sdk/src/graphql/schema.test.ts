import { GraphQLIntrospection } from '../definition';
import { cleanupSchema } from './schema';
import { buildSchema } from 'graphql';

interface TestCase {
	schema: string;
	introspection: GraphQLIntrospection;
}

const runTest = (testCase: TestCase, snapshotName: string) => {
	const schema = buildSchema(testCase.schema);
	const resultSchema = cleanupSchema(schema, testCase.introspection);
	expect(resultSchema).toMatchSnapshot();
};

describe('cleanupSchema', () => {
	describe('renamed query types', function () {
		const testCase = {
			schema: `
				schema {
					query: QueryType
					mutation: MutationType
					subscription: SubscriptionType
				}
				type QueryType {
					foo: String
				}
				type MutationType {
					foo: String
				}
				type SubscriptionType {
					foo: String
				}
				type Foo {
					queryViewer: QueryType
					mutationViewer: MutationType
					subscriptionViewer: SubscriptionType
				}
			`,
			introspection: {
				url: 'http://dummy',
			},
		};

		it('should rename query type', function () {
			runTest(testCase, 'renamed query type');
		});
	});

	describe('custom Int/Float scalars', function () {
		it('should replace int scalar', function () {
			const testCase = {
				schema: `
					scalar CustomInt
					type Query {
						foo: CustomInt
					}
				`,
				introspection: {
					url: 'http://dummy',
					customIntScalars: ['CustomInt'],
				},
			};

			runTest(testCase, 'custom int scalar');
		});

		it('should replace float scalar', function () {
			const testCase = {
				schema: `
					scalar CustomFloat
					type Query {
						foo: CustomFloat
					}
					`,
				introspection: {
					url: 'http://dummy',
					customFloatScalars: ['CustomFloat'],
				},
			};

			runTest(testCase, 'custom float scalar');
		});
	});
});
