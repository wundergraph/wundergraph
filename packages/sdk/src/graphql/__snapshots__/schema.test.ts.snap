// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cleanupSchema renamed query types should rename query type 1`] = `
"schema {
  query: Query
  mutation: Mutation
  subscription: Subscription
}

type Query {
  foo: String
}

type Mutation {
  foo: String
}

type Subscription {
  foo: String
}

type Foo {
  queryViewer: Query
  mutationViewer: Mutation
  subscriptionViewer: Subscription
}"
`;
